### ISSUE TITLE
Deprecate .extend in favour of only styled(Component)

### ISSUE THREAD

**Issue by pke** [styled-components/styled-components#1546]
The FAQ explains the difference between `styled()` and `.extend`. However I wonder if it would be possible to take the decision out of the devs hand and have `styled` check for the handed in Component if its already a styled component. If it is `.extend` it otherwise create a new class.

Was this considered at one point?

**Comment by kitten** [comment]
@mxstbr Totally agreed! We've discussed removing it before—and there was some general concensus—so I think it's time to pull the trigger.

**Comment by kitten** [comment]
@mxstbr Do you have any reason to believe that we need to keep it around for another major version? 😅 I think that won't actually make a huge difference

**Comment by mxstbr** [comment]
I don't think folks will be particularly happy about us suddenly throwing warnings in a minor version upgrade.

**Comment by pke** [comment]
Thanks for considering this simplification of the API surface. I agree such removal of API should not happen in a minor update. Please stay true to semver.
The behaviour of `styled` can be changed in a minor version but the deprecation warning could be introduced in the next major RC.

On the other hand why not release a new major version right away. What to wait for?

**Comment by mxstbr** [comment]
Yeah exactly.

**Comment by kitten** [comment]
Yea, I don't have a problem with issuing a major version asap. However, semver doesn't restrict deprecations to major releases. That'd be quite extreme as you'd then during a typical release flow have to release two major bumps—one potentially without breaking changes—just to get new APIs in and remove old ones.

Quote:

> When you deprecate part of your public API, you should do two things: (1) update your documentation to let users know about the change, (2) issue **a new minor release with the deprecation in place**. Before you completely remove the functionality in a new major release there should be at least one minor release that contains the deprecation so that users can smoothly transition to the new API.

https://semver.org/

But anyhow, I'm fine with both, but I'd like to coordinate a new release with #1493 and all related deprecations (`injectGlobal`). So I'd love to push out a major only when #1493 is done, and deprecate in a minor, but I'll leave the call up to you, @mxstbr :+1:

Btw I'll rename the issue to represent the new plan.

**Comment by mxstbr** [comment]
Yep, sounds very reasonable to batch that together with #1493! Maybe also #1171? I'd love to land that, personally.

**Comment by probablyup** [comment]
We had a similar convo last year: https://github.com/styled-components/styled-components/issues/1087

I'm definitely in favor of getting rid of `.extend` from the public API.

**Comment by kitten** [comment]
@probablyup @mxstbr to reiterate a little, I'd wait until both aforementioned PRs are merged and available at which point we can safely deprecate `.extend`, `injectGlobal`, and `keyframes` (if an alternative will be available) in a minor. Then in the next major we can remove them.

But we could already start a pending PR that adds the deprecation warnings I suppose?

**Comment by mxstbr** [comment]
Sounds perfect

**Comment by pke** [comment]
Reading #1087 I see back then there were strong arguments by members of the project against flattening the API. Good to see decisions can be reverted if they seem to confuse to many users (!) of the library even if the owners initially thought broadening the API surface was a good thing.

**Comment by kitten** [comment]
@pke to add some more depth to this prior discussion, what @probablyup was seeing became an increasingly large problem on the implementation side—kudos for seeing this so early btw—and while we preferred the API, over time it became clear that we didn’t even agree ourselves on its practical usage and use cases. Together with having problems with it (component IDs), and having no technical reason for it (specificity correctness is guaranteed, so why concat css?) this is proving to be the way forward.

Hope this dimes up everyone’s opinion on the matter :)

**Comment by tysonmatanich** [comment]
If I am following this correctly, moving forward `styled(Fruit)` would work how `Fruit.extend` does today?

```jsx
const Fruit = styled('div')`
  color: green;
`;
const Orange = styled(Fruit)`
  background: orange;
`;
```

```jsx
const Fruit = styled('div')`
  color: green;
`;
const Orange = Fruit.extend`
  background: orange;
`;
```
I'm all for the simplification, however it would be great if the original styles didn't have to be duplicated which is the same issue I have with how `.withComponent()` works. Obviously in my sample it’s not a big deal. However, if `Fruit` had tons of styles and needed to be extended many times, then it seems inefficient to duplicate the styles each time instead of just having these styles defined on a single CSS class.

**Comment by denkristoffer** [comment]
How would you style a styled component created with `withComponent`? Currently I sometimes do
```js
const Fruit = styled.div``

const Orange = Fruit.withComponent('span').extend``
```

Will the following work instead?
```js
const Orange = styled(Fruit.withComponent('span'))
```

**Comment by mxstbr** [comment]
@denkristoffer yes.

**Comment by pungggi** [comment]
hi, just to add my 2 cents.. I totally agree, this is indeed confusing when starting with styled-components. so i like the idea, as a consumer, not to think about what kind of components I am extending, but here it comes...

The documentation stated for a while i think? that `.extend` is the reccomended way.. but now the 'recommended' way will be deprecated?
Why not just leave..**both**.. so that it it doesnt matter wichone you choose. So, eliminating, the confusing but leave the choice (personal flavour i would say) to use `styled(Component)` or `Component.extend`. Personally i like the second version, because its natural to start thinking wich Component is use, and then what i want to do with it (thank you intellisense..)..

**Comment by ckeeney** [comment]
If `extend` is deprecated and removed, will there by an analogous `.attrs()` for `styled`?

**Comment by denkristoffer** [comment]
@ckeeney `attrs()` works with `styled()` as it is:

```js
const A = styled(B).attrs({
  style: ({ opacity }) => ({
    opacity,
  }),
})`
  color: red;
`

**Comment by ckeeney** [comment]
That's great news.  Thanks, I must have missed that in the docs and thought
it only worked on extend

**Comment by ms88privat** [comment]
We should be careful here. `extend` has not the same behaviour as `styled`. In our case we grep the component displayName in our visual tests. There it makes a difference what we use. If we use extend, it will not get a new displayName, but with styled it does. E.g.:

![image](https://user-images.githubusercontent.com/6705521/39112218-1cfd01c2-46d8-11e8-8a52-f1b0c11816c5.png)

Will result in:

![image](https://user-images.githubusercontent.com/6705521/39112235-2743544c-46d8-11e8-9eb0-ff6bb05b29f4.png)

**Comment by evan-scott-zocdoc** [comment]
@kizu a new stable classname is created for each individual styled component, have no fear

**Comment by kizu** [comment]
The fear is that `& > ${Button}` inside styles of some other `Foo` component won't target a `NewButton` component.

**Comment by evan-scott-zocdoc** [comment]
For that case, I'd recommend tagging the base component with some well known class or attribute and referring to that. We don't formally support inheritance in component selectors.

e.g.

const BUTTON_CLS = 'button';

```jsx
const Button = styled.button.attrs({ className: BUTTON_CLS })`
  color: red;
`;

const OtherButton = styled(Button)`
  color: blue;
`;

const Wrapper = styled.div`
  .${BUTTON_CLS} {
    background: green;
  }
`;

<Wrapper>
  <Button>Hi</Button>
  <OtherButton>Hello</OtherButton>
</Wrapper>

```

Both would have green backgrounds.

**Comment by kizu** [comment]
Sorry, but there is literally a section dedicated to this method in the documentation: https://www.styled-components.com/docs/advanced#referring-to-other-components, so it is strange to hear that this is something which is not “formally supported”.

As this is already in the doc, and people already use this method which documentation calls as allowing for “extremely powerful composition patterns”, this could be a really big hit to a lot of projects.

**Comment by evan-scott-zocdoc** [comment]
There is nothing in that document about inheritance. We support component
selectors on specific styled-components. If you make a new styled component
off another one, the second is not targeted by a component selector on the
first.


On May 25, 2018 at 2:56:40 PM, Roman Komarov (notifications@github.com)
wrote:

Sorry, but there is literally a section dedicated to this method in the
documentation:
https://www.styled-components.com/docs/advanced#referring-to-other-components,
so it is strange to hear that this is something which is not “formally
supported”.

As this is already in the doc, and people already use this method which
documentation calls as allowing for “extremely powerful composition
patterns”, this could be a really big hit to a lot of projects.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/styled-components/styled-components/issues/1546#issuecomment-392150141>,
or mute the thread

**Comment by kizu** [comment]
Ok, I understand what you're talking about, but still — right now this works this way, there are projects that rely on this feature and, yes, 4.0.0 could be breaking in this regards, but without a built-in escape hatch developers would need to cheat and make it so styled-components won't understand that you're passing an existing styled component to `styled()`. No one would change the existing behavior that just works and does the work perfectly to some classes that are not guaranteed to be unique.

Mixing multiple classes on one element is a really powerful feature and it is strange that it is not supported. Ok, it obviously has its drawbacks, and that is why I'm ok with the default behavior to be the same as extend.

**Comment by evan-scott-zocdoc** [comment]
> Mixing multiple classes on one element is a really powerful feature and it is strange that it is not supported

What do you mean? styled-components let's you do whatever you want. It just generates a stable unique className per styled component class and appends it to the rendered DOM. That's what is used when composing a component selector.

**Comment by evan-scott-zocdoc** [comment]
The difference between `Thing.extend` and `styled(Thing)` is that `Thing.extend` makes a copy of all the styles of `Thing` and assigns a single new dynamic `className` for them.

If you're referring in code to the dynamic classNames generated by a styled-component and not the stable one (`Thing.styledComponentId`), I highly recommend avoiding that as it isn't part of the API of this library and could change at any time.

**Comment by tysonmatanich** [comment]
However this duplicates the styles so they are defined on separate class names instead of either sharing a single class name or having a selector that contains multiple class names (`.a, .b { }`). Has anyone studied the performance impact this has for both client and server rendered sites?

**Comment by kizu** [comment]
Yes, I'm acutely aware of how `extend` and `styled()` work. I'm not referring in code to the dynamic classNames, and I have already written that I use it per the docs I have mentioned — by using `${MyComponent}` as a part of the selector.

The problem, again, is that the change in 4.0.0 makes styled-components to treat all other styled-components in a special way from any other components. It would dismiss the original components' stable className.

There are cases when we **need** both pairs of classNames: those from the original component, and the new ones. We need both stable classNames on one element.

I fell like I'm repeating myself.

**Comment by probablyup** [comment]
This API has been removed in the v4.0 branch per the [roadmap](https://github.com/styled-components/styled-components/issues/1694). We'll add a deprecation message to v3.x in an upcoming minor release.

**Comment by sdalonzo** [comment]
Could the docs get updated too? I've been following https://www.styled-components.com/docs/faqs#which-one-should-you-use and have created unnecessary technical debt for my team.

**Comment by mxstbr** [comment]
@sdalonzo PR to the docs would be much welcome: https://github.com/styled-components/styled-components-website

**Comment by elektronik2k5** [comment]
I love the work that's done here, and deprecation warnings are important, but:
1. The new warning is really spamming Jest test output. Can we somehow show it only once? I think that's how it works in the browser, but not in Jest.
1. Why use `console.error` for a warning? Is it for the stack trace (`console.warn` shows it too)? Is it for increased visibility? I believe those aren't good reasons. It should be a warning.

Thanks again! :+1:

**Comment by probablyup** [comment]
@elektronik2k5 I think it's because Jest resets the module registry regularly and when it requires the file fresh, the "once" shortcircuit gets reset since the whole function is recreated. Switching it to a warning sounds fine though.

**Comment by kachkaev** [comment]
@Sly777 there is at lease one use case to watch out for if I remember things correctly:

```js
const X2 = X.extend`...`.withComponent('div');
```

should become

```js
const X2 = styled(X.withComponent('div'))`...`;
```

https://github.com/styled-components/styled-components/issues/1880#issuecomment-414078886

**Comment by Sly777** [comment]
Aaa I see. I didn't know that usage way.

**Comment by RIP21** [comment]
@mxstbr is there is a need to write a codemod? I can write one with pleasure (babel-codemod) :)

**Comment by probablyup** [comment]
@RIP21 please feel free! We'd then want to add a link to it in the FAQ section of the website

**Comment by sbrichardson** [comment]
If not mentioned, the codemod would also need to add the import to any files?

```js
import styled from 'styled-components'
```

**Comment by RIP21** [comment]
Yes, I'm already testing it and this thing is supported.
Also if `{ css }` is there `styled` will be appended to the left as default
import.

**Comment by RIP21** [comment]
So, codemod is here 🎉
https://github.com/RIP21/styled-components-v3-to-v4-codemod
Check this thing out.

The only thing I can't publish on `npm` under this name for some stupid anti-spam policies reasons. 
I already contacted `npm` to get is sorted out. Feel free to take `src/index.js` file and test it on your codebase following instructions and common sense :)

**Comment by RIP21** [comment]
@Osuriel if you want to extend the styles of any component just wrap it in styled function. `const RedComponent = styled(Component)`color: red;` `

Extend is history now.

**Comment by ppozniak** [comment]
Maybe anyone can explain this issue to me:
```
const Button = styled.button`
 // ... All basic styling
`

const CTAButton = styled(Button)`
  //... Just override a few colors, etc.
`

const CTAButtonLink = CTAButton.withComponent('a');
```

And now my `CTAButtonLink` has only `CTAButton` styling, and lost all the `Button` styling.
With `const CTAButton = Button.extend` everything worked fine. What I'm doing wrong here and how could I fix that?
Thanks.

**Comment by RIP21** [comment]
@ppozniak looks absolutely valid.
Try to reproduce that in codesandbox.io and open issue if it's the case.

**Comment by probablyup** [comment]
I commented in the issue thread. The behavior is valid, just a misunderstanding of how the API works.

**Comment by greypants** [comment]
**Is it still the plan to have `styled(Component)` detect if `Component` is a styled-component or not, and if so, apply the `.extend` logic under the hood?**

We're getting the deprecation warnings, but we can't actually upgrade until the `.extend` functionality is restored under the hood of `styled()`, as was suggested in the original description of this issue. 

Current compatibility issues with `.withComponent` are highlighted below:

```js
const Button = styled.button`
  // Imagine some styles here
`;

const ExtendedButton = Button.extend`
  color: green;
`;

const StyledButton = styled(Button)`
  color: green;
`;

// FAILS :(
const StyledButtonLink = styled(StyledButton.withComponent('a'))`
  font-weight: bold;
`
// PASSES :) (because we used .extend)
const ExtendedButtonLink = styled(ExtendedButton.withComponent('a'))`
  font-weight: bold;
`
```

Here's a code pen demonstrating where this breaks down, and why we can't upgrade our app:

https://codepen.io/greypants/pen/KxaBvx?editors=0010

**Comment by probablyup** [comment]
Hey @greypants, could you see the convo here? https://github.com/styled-components/styled-components/issues/1956

If you're using `withComponent` for the purpose of switching out what physical element is being rendered, a polymorphic component base makes much more sense.

**Comment by greypants** [comment]
@probablyup woaaaaaaah... that pattern is awesome, thanks! 

So the answer then, is "No"? The `.extend` functionality is being removed completely, not moved back into `styled()`.

 Does that mean `.withComponent` should also be deprecated in favor of the polymorphic component pattern, since it's incompatible with `styled()`?

**Comment by probablyup** [comment]
> Does that mean .withComponent should also be deprecated in favor of the polymorphic component pattern, since it's incompatible with styled()?

We haven't made a formal decision on it yet, but probably. I'm doing some experimentation to see how difficult it would be to bring the polymorph functionality in as a first class prop API in `styled`

**Comment by theboyWhoCriedWoolf** [comment]
Hello,
It seems that without the .extend functionality, there is no attribute validation within styled-components.

So the following breaks:
```
  it('should not adopt an invalid attribute', () => {
    const renderedComponent = mount(<RepoLink attribute="foo" />);
    expect(renderedComponent.prop('attribute')).toBeUndefined();
  });
```

With the component:

```
const CompA = styled(CompB)`
  height: 100%;
  color: black;
`;
```

**Comment by mxstbr** [comment]
@RIP21 thanks so much for that codemod! I've moved it to the styled-components organization and packaged it into a single executable for nicer usage: https://github.com/styled-components/styled-components-codemods

Will publish a first beta version to npm to try it out :tada:

**Comment by RIP21** [comment]
@mxstbr glad to hear! If any issues let me know! :)

**Comment by lauriejones** [comment]
Apologies for posting here but I have tried to search through issues and the change log but would appreciate some clarification.

At what version did `styled()` get updated to deprecate `.extend`? 

Obviously in v4 `.extend` is no more and `styled()` "folds" the functionality when trying to style a styled component.

I can see in [v3.4.3](https://github.com/styled-components/styled-components/blob/master/CHANGELOG.md#v343---2018-08-21) that a warning was added when using `extend` -- does that mean at this point `styled()` is handling styling styled-components the way `extend` used to?

We are using `extend` in a number of places, in apps with a range of versions (back to v2) and am unsure about when I can swap out our `.extend`s in our shared components.

Thanks!

**Comment by RIP21** [comment]
AFAIK styled behaves the same way as it did before.
Follow the instructions in codemods repo here in this organization to find
out the tiny difference. Long story short it's dangerous when used in
conjunction with "withComponent" otherwise quite the same behavior.

24 paź 2018 06:54 "Laurie Jones" <notifications@github.com> napisał(a):

Apologies for posting here but I have tried to search through issues and
the change log but would appreciate some clarification.

At what version did styled() get updated to deprecate .extend?

Obviously in v4 .extend is no more and styled() "folds" the functionality
when trying to style a styled component.

I can see in v3.4.3
<https://github.com/styled-components/styled-components/blob/master/CHANGELOG.md#v343---2018-08-21>
that a warning was added when using extend -- does that mean at this point
styled() is handling styling styled-components the way extend used to?

We are using extend in a number of places, in apps with a range of versions
(back to v2) and am unsure about when I can swap out our .extends in our
shared components.

Thanks!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/styled-components/styled-components/issues/1546#issuecomment-432512429>,
or mute the thread

**Comment by mxstbr** [comment]
@lauriejones no, the refactor that "folds"/merges styles into each other was only released in v4, although I'm pretty sure your code should still work in v3 if changed to `styled(Comp)` from `.extend`.

**Comment by mmcc** [comment]
I believe I'm running into the same issue that @lauriejones described. After upgrading and moving over to `styled` in v4, I'm running into specificity issues all over the place. It's hard to tell if it's because something `extend` was doing was masking actual issues, but it's certainly frustrating to debug.

**Comment by mimi4Real** [comment]
> **Is it still the plan to have `styled(Component)` detect if `Component` is a styled-component**, and if so, apply the `.extend` logic under the hood?
> 
> We're getting the deprecation warnings, but we can't actually upgrade until the `.extend` functionality is restored under the hood of `styled()`, as was suggested in the original description of this issue.
> 
> Current compatibility issues with `.withComponent` are highlighted below:
> 
> ```js
> const Button = styled.button`
>   // Imagine some styles here
> `;
> 
> const ExtendedButton = Button.extend`
>   color: green;
> `;
> 
> const StyledButton = styled(Button)`
>   color: green;
> `;
> 
> // FAILS :(
> const StyledButtonLink = styled(StyledButton.withComponent('a'))`
>   font-weight: bold;
> `
> // PASSES :) (because we used .extend)
> const ExtendedButtonLink = styled(ExtendedButton.withComponent('a'))`
>   font-weight: bold;
> `
> ```
> 
> Here's a code pen demonstrating where this breaks down, and why we can't upgrade our app:
> 
> https://codepen.io/greypants/pen/KxaBvx?editors=0010

use `as` replace `withComponent`
